rfc/rfc-0000-allow-resources-to-report-status.md

```diff
@@ -0,0 +1,335 @@
+
+Each owner will display a new condition in the status called `ResourcesSuceeded` which will only be updated to `True` when all resources in the blueprint be healthy. Cartographer will not update the top level state of an owner to `Ready: True` until `ResourcesHealthy` is `True`.
+
+# Migration
+[migration]: #migration
+
+Given that Cartographer will default to `Healthy` once the resource is applied and outputs are read, this will not be a breaking change.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+- The owner status will potentially grow significantly, based on how many matchers are in a `ClusterResourceStatusRule`.
+- We only allow for reflecting three states, though this design allows for creating more states in the future.
+- This design does not allow authors to define the states themselves.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+  - The `ClusterResourceStatusRule` could instead be inlined in the templates.
+    - Separating these rules in a separate CRD allows for re-usability of those rules.
+  - We could also not require the templates to specify the `ClusterResourceStatusRule` and instead have those rules automatically select templates based on the `kind` and `apiVersion` of the resource they are stamping out.
+    - This could be a bit confusing because the type of the stamped resource may be templated as well. Therefore, a template could end up stamping out different types of resources and so Cartographer would only be able to inform users which `ClusterResourceStatusRule` is used for a given template after an owner is applied, and it could then be reflected in the owner status.
+- What is the impact of not doing this?
+  - Users/UI creators will have to have in depth knowledge of the underlying resources to determine the health of the owner.
+
+# Prior Art
+[prior-art]: #prior-art
+
+N/A
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Should the two states be `Healthy` and `Unhealthy`?
+- Is it right for Cartographer to default to `Healthy`?
+
+# Spec. Changes
+[spec-changes]: #spec-changes
+- All templates will now have an optional `resourceStatusRuleRef` which will take a `name`. In the future, it could also take a type and namespace if we introduce namespace-scoped resourceStatusRules.
+- Owner `status.resources` will now have `stampedStatus` on each resource:
+```yaml
+status:
+  resources:
+    - name: resource1
+      ...
+      stampedStatus:
+        status: <[Healthy|Unhealthy|Unknown]>
+        conditions:
+          - <metav1.Condition>
+        fields:
+          - key: <string>
+            value: <string>
+            message: <string>, omitempty
```

Comment:
Can we unify these structures so that users don't need to do that work themselves?

A rough idea would be to surface conditions for each resource

```yaml
status:
  resources:
  - name: resource1
    conditions: <metav1.Condition>
    ...
```

These conditions are not a copy of the stamped resource's status conditions, but a set of conditions specific to the resource as viewed by Cartographer. For the purposes of this RFC, we're concerned about a `Heathy` condition summarizing the health of the stamped resource, other condition can be introduced to surface other failure points for a particular resource, like an error applying the template, or an api server error on create/update.

In practice, this would look like:

```yaml
status:
  resources:
  - name: resource-condition
    conditions:
    - type: Healthy
      status: 'True'
      reason: ReadyCondition
      message: ...copied from the stamped resource's Ready condition's message...
      lastTransitionTime: 2022-03-17T00:29:04Z # managed by cartographer
  - name: resource-field
    conditions:
    - type: Healthy
      status: 'False'
      reason: MatchedField
      message: ...defined by the rule's message path...
      lastTransitionTime: 2022-03-17T00:29:04Z # managed by cartographer
```